ghmirror/core/mirror_requests.py

```diff
@@ -58,6 +58,70 @@ def _cache_response(resp, cache, cache_key):
         cache[cache_key] = resp
 
 
+def _online_request(method, url, cached_response,
+                    headers=None, parameters=None):
+    """
+    Handle API errors on conditional requests and try
+    to serve contents from cache
+    """
+    try:
+        resp = requests.request(method=method,
+                                url=url,
+                                headers=headers,
+                                timeout=REQUESTS_TIMEOUT,
+                                params=parameters)
+
+        # When we hit the API limit, or there is a problem with the API
+        # let's try to serve from cache
+        (serve_from_cache,
+         response_header) = _should_error_response_be_served_from_cache(resp)
+        if serve_from_cache:
+            if cached_response is None:
+                LOG.info('%s GET CACHE_MISS %s', response_header, url)
+                resp.headers['X-Cache'] = response_header + '_MISS'
+                return resp
+
+            LOG.info('%s GET CACHE_HIT %s', response_header, url)
+            cached_response.headers['X-Cache'] = response_header + '_HIT'
```

Comment:
We will never reach to here with the empty string as the function is returning a bolean along with the header, and we only get to here if `serve_from_cache` is True
```
(serve_from_cache, response_header) = _should_error_response_be_served_from_cache(resp)
if serve_from_cache:
  XXX
```

https://github.com/app-sre/github-mirror/pull/63/files/7ffac9e2f66280bc206d0eb29b1668dc3a4ab650#diff-b65064d2dfc0b574509b9afa3a22ef4675bdbe4f2ff285de9b5ce1288c7faaecR76-R77

If it's false and the string is none, we return a cache miss directly 

https://github.com/app-sre/github-mirror/pull/63/files/7ffac9e2f66280bc206d0eb29b1668dc3a4ab650#diff-b65064d2dfc0b574509b9afa3a22ef4675bdbe4f2ff285de9b5ce1288c7faaecR88-R90

But maybe is not the best way to do it (?)

Reply:
So you are using an empty string for failing the "truth value test", probably because you care about types, so it might look weird not to return a string there, but python doesn't care about types. You can safelly return `None` instead of an empty string, so we know we have to do the right thing to it, otherwise the code will break.

That's just a question of how we use the language, almost a style thing, but as long as you're doing the right thing, which you are, you don't have to do as I say :)

Reply:
Anyway, what I'd do: `_should_error_response_be_served_from_cache` would return either a string or None. If it has a string, to use it, if it is None, you go away :)

Reply:
Something like:
```python
    serve_from_cache_header = get_serve_from_cache_header()
    
    # early checkout :)
    if not serve_from_cache_header:
        return resp
    
    # process cached_response inside the "if" testing it
    if cached_response is not None:
        LOG.info('%s GET CACHE_HIT %s', response_header, url)
        cached_response.headers['X-Cache'] = serve_from_cache_header + '_HIT'
        return cached_response
    
    # last resort needs no extra processing
    LOG.info('%s GET CACHE_MISS %s', response_header, url)
    resp.headers['X-Cache'] = serve_from_cache_header + '_MISS'
    return resp
```